crates/bevy_ecs/src/signal/mod.rs

```diff
@@ -0,0 +1,203 @@
+    let setup_info = world
+        .entity_mut(subscriber_entity)
+        .take::<SubscribedComponentSetup<S, C>>()
+        .unwrap();
+
+    // Get the signal entity
+    let signal_entity = world
+        .get_entity(setup_info.signal.entity)
+        .expect("Tried to subscribe an entity's component to a signal that was despawned.");
+
+    // Get the value the signal entity is holding
+    let signal_value = signal_entity.get::<S>().expect(
+        "Tried to subscribe an entity's component to a signal that had its component removed.",
+    );
+
+    // Call the subscription function to produce the initial subscribed component value
+    let signaled_component_initial_value = (setup_info.f)(signal_value);
+
+    // Add the initial subscription value to the subscriber entity
+    world
+        .entity_mut(subscriber_entity)
+        .insert(signaled_component_initial_value);
+
+    // Setup an ongoing subscription to update the subscriber entity's component value whenever the signal value changes
+    let f = setup_info.f;
+    let subscription = move |signal_trigger: Trigger<OnMutate, S>,
+                             signal_query: Query<&S>,
+                             mut subscriber_query: Query<&mut C>| {
+        // Get the current value of the signal
+        let signal_value = signal_query
+            .get(signal_trigger.entity())
+            .expect("Signal component was removed");
+
+        // Call the subscription function to produce the subscribed component value
+        let signaled_component_value = (f)(signal_value);
+
+        // Update the subscribed entity's component value
+        *subscriber_query
+            .get_mut(subscriber_entity)
+            .expect("Subscribed component was removed") = signaled_component_value;
+    };
+    world.spawn((
+        SignalSubscriptionMarker,
+        setup_info.signal.refcount,
+        Observer::new(subscription).with_entity(setup_info.signal.entity),
+    ));
+}
+
+/// Marker component for an entity backing a [`Signal`].
+#[derive(Reflect)]
+pub struct SignalMarker;
+impl Component for SignalMarker {
+    const STORAGE_TYPE: StorageType = StorageType::Table;
+}
+
+/// Marker component for an entity backing a [`Signal`] subscription.
+#[derive(Reflect)]
+pub struct SignalSubscriptionMarker;
+impl Component for SignalSubscriptionMarker {
+    const STORAGE_TYPE: StorageType = StorageType::Table;
+}
+
+/// Refcount for tracking the lifetime of a [`Signal`].
+#[derive(Reflect, Clone, Default)]
+#[reflect(opaque)]
+pub struct SignalRefcount(Arc<()>);
+impl Component for SignalRefcount {
+    const STORAGE_TYPE: StorageType = StorageType::Table;
+}
+
+/// System to automatically clean up unused [`Signal`]s.
+pub fn signal_cleanup(signals: Query<(Entity, &SignalRefcount)>, mut commands: Commands) {
+    for (signal, refcount) in &signals {
+        if Arc::strong_count(&refcount.0) == 1 {
```

Comment:
I recommend doing what Asset handles do and have an `Arc<T>` where T queues an event to despawn the entity when the Arc data is dropped. `SignalRefcount` would probably want to hold a `Weak<T>` then. Then you don't have to iterate through all signals, just ones that actually need to be cleaned up.

Reply:
> Unless you actually need to check the reference count for other reasons, you probably don't need SignalRefcount at all with the above.

Not sure I get how, but I would love to get rid of it. Currently it's used for two things:
* If an entity (X) has a component subscribed to a signal, that's handled by spawning a new observer entity. Said observer entity gets a SignalRefcount component to ensure that the signal entity is not despawned while the observer entity is alive, i.e. while entity X has a subscription to the signal.
* If you have a Signal<T> type (which is not a bevy component or resource), it internally holds a SignalRefcount so that the signal entity is not despawned while Signal<T> is alive. You can pass this Signal<T> into observer callbacks for button clicks or event handling systems or whatever in order to call signal.set(v) to update the signal's value.

I would love to get rid of the refcount so that Signal<T> can be Copy again, but then I don't know how to automatically decide when to cleanup the signal entity.